src/python/pants/core/target_types.py

```diff
@@ -25,6 +44,129 @@ class Files(Target):
     core_fields = (*COMMON_TARGET_FIELDS, Dependencies, FilesSources)
 
 
+# -----------------------------------------------------------------------------------------------
+# `relocated_files` target
+# -----------------------------------------------------------------------------------------------
+
+
+class RelocatedFilesSources(Sources):
+    # We solely register this field for codegen to work.
+    alias = "_sources"
+    required = False
+    expected_num_files = 0
+
+
+class RelocatedFilesOriginalTargets(StringSequenceField):
```

Comment:
True. 

And maybe the field name should be `dependencies`? My rationale for the name `files_targets` is to make it more clear that this should (probably) only be `files()` in the value, not `python_library()` targets etc. Although, we can still subclass `Dependencies` and keep the alias the same.

Reply:
It's also possibly to enforce that the contents all produce `FilesSources`, which this should probably do anyway.

I think that this gets back into a general philosophical discussion similar to the "goals refer to a particular output type" discussion, but for fields: i.e. should `*_test` targets have one field per dependency output type they are expecting? etc.

Reply:
I've been thinking about this one a bit today, mostly from the lens of consistency. Users have expressed confusion already around things like `files` vs. `resources`, and I think one way we reduce confusion is to ensure things are always consistent.

> i.e. should *_test targets have one Dependencies field per dependency output type that they are expecting? etc.

Generally, I like having the `dependencies` field for everything, if those dependencies get treated how we normally handle dependencies: that we copy their `sources` files into the chroot and pull in their own deps. I think it would be clunky and confusing to have `files_dependencies`, `resources_dependencies`, and `python_library_dependencies`, etc.

But, I think that any dependencies which we do not treat the "normal" way and we do something special, it should not be in the `dependencies` field. 

For example, normally depending on a target would not trigger some big side effect like running the equivalent of `./pants binary`. This is why we added `runtime_binary_dependencies`; we want to call special attention to it. Importantly, we do not enforce that a `python_library` cannot depend on a `python_binary`, and even dep inference can trigger that dependency if you are importing the source file associated with the binary. There, the dep is like any other dep, you just get the source file. So, we maintain consistency by avoiding the case where a dep on `python_binary` sometimes mean to compile it, and sometimes to just use the source file.

We could invent mechanisms to ban dependencies on certain target types, but that might prove tricky to implement with plugins. If a user writes a `pyinstaller_binary` target, our prebuilt code wouldn't know to ban a dep on it.

--

This general heuristic would I think suggest this behavior:

* Here, we should use the name `files_targets`, rather than `dependencies`. We are doing something magical of transforming the target's `dependencies` into the `sources` of this target via codegen. That's not normal.
* With `archive`, we should use `assets`, as we're doing something magical of calling `./pants package/binary`. We likely want to have `files` as well; that's not doing anything magical, though, so that part isn't as clear to me.

Reply:
Oh, I don't think that we can change this field to subclass `Dependencies`. That would result in `TransitiveTargets` pulling in the original files targets, which means that the unmodified sources files might still show up in your chroot; it's important that we do not do that. I'll add tests for that edge case, I could see it biting us in the future.

But it would be nice for these dependencies to still show up in project introspection code. We might want to invent a new mechanism that distinguishes between normal dependencies vs. special dependencies like this, such that project introspection code will look at the special deps too, but normal code wouldn't.